### PR TITLE
Better handle missing tokens when formatting so that we can respect options like "use tabs".

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes: https://github.com/dotnet/roslyn/issues/4014

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
reviewers: @balajikris, @basoundr, @brettfo, @davkean, @DustinCampbell, @dpoeschl, @Pilchie, @jasonmalinowski, @heejaechang

**Comment by DustinCampbell** [comment]
Nice fix. :+1:

**Comment by heejaechang** [comment]
:+1:

File: src/Workspaces/Core/Portable/Formatting/Engine/AbstractTriviaDataFactory.AbstractComplexTrivia.cs [hunk 1]
```diff
@@ -41,54 +41,21 @@ protected abstract class AbstractComplexTrivia<TToken, TTrivia> : TriviaDataWith
...
             protected abstract bool ContainsSkippedTokensOrText(TriviaList list);
 
-            public TToken Token1
-            {
-                get
-                {
-                    return ConvertToken(_token1);
-                }
-            }
+            public SyntaxToken Token1 => _token1;
```
Comment (Pilchie) [inline src/Workspaces/Core/Portable/Formatting/Engine/AbstractTriviaDataFactory.AbstractComplexTrivia.cs]: Couldn't you go further and make this `public SyntaxToken { get; }`?  Or is _token not assigned in the constructor?

**Comment by CyrusNajmabadi** [comment]
Baby steps :)   I'd like to actually write a analyzer/fixer for that.

File: src/Workspaces/Core/Portable/Formatting/Engine/AbstractTriviaDataFactory.AbstractComplexTrivia.cs (hunk 1 above)
Comment (Pilchie) [inline src/Workspaces/Core/Portable/Formatting/Engine/AbstractTriviaDataFactory.AbstractComplexTrivia.cs]: Have you seen https://github.com/DustinCampbell/CSharpEssentials?

**Pull request merged**